### PR TITLE
[fix] `static-property-placement`: fix bug with static property refs inside methods

### LINKED ISSUE #2283: [static-property-placement] Violation on accessing static variable
Example:
```typescript
class MyComponent extends React.Component<MyProps, MyState> {
	public static displayName = "MyComponent ";
	public myMethod() {
		console.log(MyComponent.displayName);
	}
}
```

In the example, line 4 is flagged as a static-property-placement violation even though it simply accesses the property.

### PR DESCRIPTION
**Pull request by dmason30**
Fixes #2283

### CODE REVIEW

File: lib/rules/static-property-placement.js [hunk 1]
```diff
@@ -129,9 +129,16 @@ module.exports = {
       ClassProperty: node => reportNodeIncorrectlyPositioned(node, STATIC_PUBLIC_FIELD),
 
       MemberExpression: node => {
+        // If definition type is undefined then it must not be a defining expression or if the definition is inside a
+        // class body then skip this node.
+        if ((node.parent.right || {}).type === 'undefined' || isContextInClass()) {
```
Comment (ljharb) [inline lib/rules/static-property-placement.js]:
```suggestion
        const { right } = node.parent;
        if (!right || right.type === 'undefined' || isContextInClass()) {
```

File: tests/lib/rules/static-property-placement.js [hunk 2]
```diff
@@ -1072,6 +1072,35 @@ ruleTester.run('static-property-placement', rule, {
           }
         }
       `].join('\n')
+    },
+    // ------------------------------------------------------------------------------
+    // edge cases
+    // ------------------------------------------------------------------------------
+    {
+      // Do not error if property assignement inside class function
```
Comment (ljharb) [inline tests/lib/rules/static-property-placement.js]:
```suggestion
      // Do not error if property assignment is inside a class function
```

File: lib/rules/static-property-placement.js [hunk 3]
```diff
@@ -129,9 +129,17 @@ module.exports = {
       ClassProperty: node => reportNodeIncorrectlyPositioned(node, STATIC_PUBLIC_FIELD),
 
       MemberExpression: node => {
+        // If definition type is undefined then it must not be a defining expression or if the definition is inside a
+        // class body then skip this node.
+        const { right } = node.parent;
```
Comment (ljharb) [inline lib/rules/static-property-placement.js]: whoops, looks like this breaks in node 4? if so:
```suggestion
        const right = node.parent.right;
```

File: lib/rules/static-property-placement.js [hunk 4]
```diff
@@ -86,7 +86,7 @@ module.exports = {
      * Checks if we are declaring function in class
```
Comment (jp7837) [inline lib/rules/static-property-placement.js]: update?
```suggestion
      * Checks if we are declaring context in class
```

File: lib/rules/static-property-placement.js [hunk 5]
```diff
@@ -86,7 +86,7 @@ module.exports = {
...
      * @returns {Boolean} True if we are declaring function in class, false if not.
```
Comment (jp7837) [inline lib/rules/static-property-placement.js]:
```suggestion
      * @returns {Boolean} True if we are declaring context in class, false if not.
```

**Comment by jp7837** [comment]
Buddy test pass on my local code :)
